### PR TITLE
Preserve board_id across archive/unarchive

### PR DESCRIPTION
**Pull request by bdraco**
## Summary

Reported by Marcel: archiving a device dropped its ``board_id`` (and other identity fields) from the metadata sidecar, so unarchive lost the catalog → YAML binding. ``board_id`` is what we use to match the catalog board with the YAML; losing it on every archive cycle forced a re-derive (or a re-pick by the user) on unarchive that wasn't necessary.

## What changed

#105's archive flow wiped the entire device-metadata entry alongside the build dir + StorageJSON sidecar. That closed a real "future same-name device inherits stale state" risk (per-filename keying), but it was over-broad — only the *volatile* fields actually go stale on archive.

Split the metadata wipe by concern:

| Field | On archive | Why |
|---|---|---|
| ``ip`` | cleared | last-resolved IP, meaningless once the device is no longer on the network |
| ``expected_config_hash`` | cleared | last-compiled hash; build dir is wiped so the hash no longer corresponds to any artifact |
| ``board_id`` | **kept** | catalog → YAML match key, derived from the YAML — same YAML on unarchive ⇒ same answer |
| ``friendly_name`` | **kept** | user label, lives in the sidecar (ESPHome doesn't track it) |
| ``comment`` | **kept** | same as friendly_name |

If the entry winds up empty after the volatile-fields clear (no identity fields ever set) it's dropped entirely so the metadata file doesn't accumulate stale keys.

The StorageJSON sidecar (``firmware_bin_path`` / ``loaded_integrations`` / ``target_platform``) is still wiped — it's a pure build artifact and would mislead any same-name new device.

New helpers:

- ``clear_volatile_device_metadata(config_dir, filename)`` in ``controllers/config.py`` — pops the volatile fields under one ``metadata_transaction``.
- ``_archive_clear_device_sidecars`` in ``controllers/devices/helpers.py`` — variant of ``_remove_device_sidecars`` that keeps identity fields. ``delete`` / ``delete_archived`` keep using the original full-removal helper.

The remaining same-name-new-device leakage of identity fields is acceptable: ``board_id`` is re-derived (or supplied) by the new device's create flow; ``friendly_name`` / ``comment`` are user labels the editor can overwrite.

## Test plan

- [x] ``uv run pytest tests/controllers/devices/test_archive.py`` — 39 tests, includes the replacement ``test_archive_clears_volatile_metadata_keeps_identity`` (asserts the exact post-archive shape) and ``test_archive_drops_metadata_entry_when_only_volatile_fields`` (empty-after-clear edge case).
- [x] Full suite: ``uv run pytest --ignore=tests/benchmarks`` — 738 passed.

### CODE REVIEW

File: esphome_device_builder/controllers/config.py [hunk 1]
```diff
@@ -311,6 +311,69 @@ def remove_device_metadata(config_dir: Path, filename: str) -> None:
...
+    + clears volatile so unarchive restores the user-visible
+    state unchanged. Same-name new-device leakage of identity
+    fields is acceptable: the new device's create flow either
+    derives or supplies its own ``board_id``, and friendly_name
+    / comment are user labels the new device's editor can
+    overwrite if desired.
+    """
+    with metadata_transaction(config_dir) as data:
+        entry = data.get(filename)
+        if not isinstance(entry, dict):
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome_device_builder/controllers/config.py:368]: Real concern, fixed in c0802ee. The early-return on non-dict entries was indeed lossy: a corrupt value would survive the clear, then `set_device_metadata`'s item-assign into `entry` would crash. Now we drop the bad entry inside the same transaction so the next write starts from a clean shape.

Pinned by `test_clear_volatile_device_metadata_drops_corrupt_non_dict_entry`. Seeded via `_save_metadata` directly because `set_device_metadata` won't let a non-dict value through its public surface.

File: esphome_device_builder/controllers/config.py [hunk 2]
```diff
@@ -311,6 +311,69 @@ def remove_device_metadata(config_dir: Path, filename: str) -> None:
...
+    risk but also lost the identity fields. The catalog → YAML
+    match key is ``board_id``; losing it on every archive →
+    unarchive cycle forced a re-derive (or a re-pick by the
+    user) that wasn't necessary. This helper preserves identity
+    + clears volatile so unarchive restores the user-visible
+    state unchanged. Same-name new-device leakage of identity
+    fields is acceptable: the new device's create flow either
+    derives or supplies its own ``board_id``, and friendly_name
+    / comment are user labels the new device's editor can
+    overwrite if desired.
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome_device_builder/controllers/config.py:362]: Real bug, fixed in c0802ee. The stub-create path (no `board_id` provided + no derive match) never reached the `set_device_metadata` write, so an archived same-name entry's `board_id` / `friendly_name` / `comment` survived and silently mis-bound the new device. Fix: `remove_device_metadata` at the top of `_init_storage` before optionally writing the supplied/derived `board_id`. Archive still preserves identity for the unarchive path; create explicitly wipes for the same-name-new-device path.

Two new tests:
- `test_create_device_clears_residual_metadata_from_archived_same_name` — seeds the stale entry and asserts the post-create entry is `{}` (stub flow has no `board_id` to write back).
- `test_create_device_with_board_id_overwrites_archived_board_id` — explicit `board_id="rp2040-new-board"` wins over `board_id="esp32-archived-board"`; final entry is `{"board_id": "rp2040-new-board"}`.

Decided against namespacing archived entries (`archive:<filename>` keys) — the create-side wipe is one line and doesn't change the on-disk schema; namespacing would be a wider migration with no other consumer.
